### ISSUE TITLE
Docker classic, Traefik & X-Forwarded-For IP

### ISSUE THREAD

**Issue by edvincent** [acouvreur/sablier#117]
_Not sure this qualifies as a bug, or as a "known limitation" that should be highlighted.._

To the best of my knowledge (and I spent some time researching this in the past - but maybe I'm missing something that can be done which would be awesome), when using Traefik in a non-swarm (aka Docker classic) setup, the only way to keep the original requestor IP is using `network_mode: "host"`. That's because using anything else will involve Docker's NATing, which will lose the original IP (as this is done at the TCP/IP level).

This is why, in the documentation example [Quick start with Traefik](https://github.com/acouvreur/sablier#quick-start-with-traefik), we see `X-Forwarded-For: 172.18.0.1` which is the default gateway of the network the container is in - rather than the requester's IP.

The limitation comes if you want to get the original IP as well as using Sablier. To get the original IP, you can use the `host` networking mode. But the problem with that is that it breaks the "by container name/alias" resolution - as the host doesn't have DNS entries for the containers... Which breaks the setup, [as the dynamic rule relies on the DNS name ](https://github.com/acouvreur/sablier/blob/main/dynamic-config.yml#L6)(as the IP itself of the container can change...) `- url: "http://whoami:80"`.

A "workaround" that I've found (**only works for the dynamic strategy**) that works well is actually relying on Traefik's rules priority:
1. On the dynamic config (the one that stays even when the container is down), I have the "normal" rule pointing to `container.domain.com`, and a NoOp service (just pointing to a 0.0.0.0 URL) - along with the Sablier middleware.
2. On the container traefik labels, I duplicate the rule, with a "longer" rule, that forces Traefik to prefer this one if both are available, which is the case once the container is launched and traefik sees both rules. The rule is effectively `Host(container.domain.com) || Host(container.domain.com)`, which is longer, but does the same thing.

That way, whenever the browser refreshes the page, it will hit the new rule. Obviously, it's a bit painful to have to duplicate the rule, but it works. I also keep the same middleware in the rule, to make sure the session is refreshed.

This obviously doesn't work for the `blocking` strategy, as there's no refresh involved, so no opportunity for traefik to re-evaluate the rules.

**Comment by acouvreur** [comment]
Where does the issue stands in the first place?

Is this about the IP showing in the log of Sablier?

Because as a Middleware, it should not interfere with headers injected by Traefik such as X-Forwarded-For.

As for the known limitation https://github.com/moby/moby/issues/25526, it seems to affect swarm mode only?

**Comment by edvincent** [comment]
It's not a Sablier bug, but it's a limitation of setting it up the way it's recommended in the guide/example.

It affects any application that wants/needs to have the original IP of the requester when using Traefik in a non-swarm mode (e.g. to allow-list based on IPs in Traefik itself, or logging purpose anywhere downstream). If you want that, you need to use the host networking mode in the Traefik container.

And if you use Traefik in host networking mode, you cannot use Sablier (at least with the current guide), because the following config won't work:

```
http:
  services:
    whoami:
      loadBalancer:
        servers:
        - url: "http://whoami:80"
```

Because the server uses the `whoami` reference, which is based on DNS resolution, which doesn't work in host networking mode.

**Comment by acouvreur** [comment]
Intersting, maybe your setup has support for the Proxy Protocol. https://doc.traefik.io/traefik/routing/entrypoints/#proxyprotocol

Which I believe doesn't require you to expose to host, but maybe I'm wrong.

You'll also need an other reverse proxy, but layer 4 in this case... I used HAProxy with passthrough and Proxy Protocol to Traefik, worked great.

**Comment by edvincent** [comment]
It's not a limitation of Traefik either, it's just a limitation of Docker itself and how it does networking, I think? If you don't use host networking, i.e. use a docker network (which is what enables DNS resolution by container name), there's some NATing involved. So any request coming into Traefik gets NATed back to a different IP, which then regardless of what Traefik supports, the "original IP" is lost...

If you look at the output of your `whoami` page in the README example, you can see `X-Forwarded-For: 172.18.0.1`, and that `.1` is the Docker NATing at play. It's the gateway of the docker network, rather than the actual requester.

In other words, it's impossible (with your recommended setup) to have the `whoami` page show the actual IP of the requester - which is what I'm trying to do... I suspect your other setup with ProxyProtocol, HAProxy and Traefik still doesn't show the IP from the requester? Unless you're not installing Traefik with Docker in non-swarm?
